n2survey/read_from_xml.py

```diff
@@ -0,0 +1,268 @@
+import re
+import sys
+from json import dump
+from typing import Dict
+
+import bs4
+from bs4 import BeautifulSoup
+
+
+def _text_refining(s: str) -> str:
+    matched = ["\n", "\t", "\xa0"]
+    res_s = re.sub("|".join(matched), "", s)
+    return res_s
+
+
+def _sectioninfo_text_parse(bsString: str) -> str:
+    text = ""
+    sps = BeautifulSoup(bsString, "html.parser").find_all("p")
+    for sp in sps:
+        if bool(sp.string):
+            text += _text_refining(sp.string)
+    return text
+
+
+def _find_choices(resp: bs4.element.Tag) -> dict:
+    cates = dict()
+    categories = resp.find_all("category")
+    for category in categories:
+        category_lab = category.label.string
+        category_val = category.value.string
+        # A#other case
+        if bool(category.contingentQuestion):
+            category_val = category.contingentQuestion.attrs["varName"]
+            category_lab = _text_refining(
+                category.contingentQuestion.find("text").get_text()
+            )
+        cates[category_lab] = category_val
+    return cates
+
+
+def _integrate_subquestions(question: bs4.element.Tag, section_id: int) -> list:
+    res_subquestions = list()
+    subs = question.find_all("subQuestion")
+    if question.response.category is not None:
+        sub_type = "subquestion choice"
+    else:
+        sub_type = "subquestion text"
+    for sub in subs:
+        res_subquestions.append(
+            {
+                "name": sub["varName"],
+                "question": sub.find("text").get_text(),
+                "description": "",
+                "question type": sub_type,
+                "is_other": False,
+                "is_sub": True,
+                "choices": _find_choices(question.response),
+                "section_id": section_id,
+                "qustion group": question.response.attrs["varName"],
+            }
+        )
+    return res_subquestions
+
+
+def _store_questions(
+    question: bs4.element.Tag,
+    section_id: int,
+    question_content: str,
+    question_desc: str,
+) -> list:
+    """
+    question include:
+        1. multiple choice question (w/o descriptions)
+            1.1 multi choice + other choice item + "other" type question. See F3T
+            1.2 pure multi choice
+        2. single choice question (w/o descriptions)
+            2.1 single choice with "other" type question
+            2.2 pure single choice
+        3. pure "text" question
+        4. "comment" type question
+    """
+    store_question = list()
+    resps = question.find_all("response")
+
+    for resp in resps:
+        len_cate = len(resp.find_all("category"))
+        # multiple choice or contingent-> other + multiple choice (eg. "F3T")
+        if len_cate == 1:
+            if bool(resp.category.contingentQuestion):
+                is_other = True
```

Comment:
The thing is that when you have "other" then the function is supposed to return two items. What the function will return in case of "A1"?

Reply:
ye, it has been fixed in new version.